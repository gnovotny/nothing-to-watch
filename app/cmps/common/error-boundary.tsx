import React from 'react'
import config from '../../config'
import telemetry from '../../utils/telemetry/micro-sentry-telemetry'

type ErrorBoundaryState = {
  error: Error | null
}

export class ErrorBoundary extends React.Component<
  React.PropsWithChildren,
  ErrorBoundaryState
> {
  constructor(props: React.PropsWithChildren) {
    super(props)
    this.state = { error: null }
  }

  static getDerivedStateFromError(error: Error): ErrorBoundaryState {
    return { error }
  }

  componentDidCatch(error: Error, info: React.ErrorInfo) {
    console.error('ErrorBoundary caught an error', error, info)
    try {
      telemetry.captureError(error, {
        source: 'ErrorBoundary',
        componentStack: info.componentStack,
      })
    } catch {}
  }

  private copyDetails = async () => {
    try {
      const details = this.getErrorDetails()
      await navigator.clipboard.writeText(details)
    } catch {}
  }

  private dismissError = () => {
    this.setState({ error: null })
  }

  private getErrorDetails() {
    const { error } = this.state
    return [
      `message: ${error?.message ?? 'unknown'}`,
      `stack: ${error?.stack ?? 'n/a'}`,
      `userAgent: ${navigator.userAgent}`,
    ].join('\n')
  }

  render(): React.ReactNode {
    if (!this.state.error) return this.props.children

    const details = this.getErrorDetails()

    return (
      <div className='fixed inset-0 z-50 grid h-dvh w-dvw place-items-center bg-background p-6 text-foreground'>
        <div className='max-w-2xl rounded-lg border border-border bg-card p-6 shadow-sm'>
          <h1 className='mb-2 font-bold text-xl'>Something went wrong</h1>
          <p className='mb-4 text-muted-foreground text-sm'>
            The app encountered an unexpected error. This can happen if your
            browser or GPU doesnâ€™t fully support WebGL or required extensions.
          </p>

          <div className='mb-4 rounded-md border border-border bg-popover p-3 text-xs'>
            <p className='mb-1 font-semibold'>What you can try</p>
            <ul className='list-inside list-disc space-y-1'>
              <li>Reload this page</li>
              <li>Update your browser/graphics drivers</li>
              <li>Try a different browser (Chrome/Firefox)</li>
            </ul>
          </div>

          <pre className='mb-4 max-h-48 overflow-auto rounded-md border border-border bg-muted p-3 text-[11px] leading-tight'>
            {details}
          </pre>

          <div className='flex flex-wrap gap-2'>
            <button
              type='button'
              className='rounded-md border border-input bg-secondary px-3 py-2 text-sm hover:bg-secondary/80'
              onClick={() => window.location.reload()}
            >
              Reload
            </button>
            <button
              type='button'
              className='rounded-md border border-input bg-secondary px-3 py-2 text-sm hover:bg-secondary/80'
              onClick={this.copyDetails}
            >
              Copy details
            </button>
            <a
              className='rounded-md border border-input bg-secondary px-3 py-2 text-sm hover:bg-secondary/80'
              href={config.sourceCodeUrl}
              target='_blank'
              rel='noreferrer noopener'
            >
              Report issue
            </a>
            <button
              type='button'
              className='rounded-md border border-input bg-secondary px-3 py-2 text-sm hover:bg-secondary/80'
              onClick={this.dismissError}
            >
              Dismiss
            </button>
          </div>
        </div>
      </div>
    )
  }
}

export default ErrorBoundary
