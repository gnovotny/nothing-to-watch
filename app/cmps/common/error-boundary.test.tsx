import { render, screen } from '@testing-library/react'
import { describe, expect, it, vi } from 'vitest'
import { ErrorBoundary } from './error-boundary'

// Mock the telemetry module
vi.mock('../../utils/telemetry/micro-sentry-telemetry', () => ({
  default: {
    captureError: vi.fn(),
  },
}))

// Mock the config module
vi.mock('../../config', () => ({
  default: {
    sourceCodeUrl: 'https://github.com/example/repo',
  },
}))

// Component that throws an error for testing
const ThrowError = ({ shouldThrow = false }: { shouldThrow?: boolean }) => {
  if (shouldThrow) {
    throw new Error('Test error message')
  }
  return <div>No error</div>
}

describe('ErrorBoundary', () => {
  it('should render children when no error occurs', () => {
    render(
      <ErrorBoundary>
        <ThrowError shouldThrow={false} />
      </ErrorBoundary>,
    )

    expect(screen.getByText('No error')).toBeInTheDocument()
  })

  it('should render error UI when child component throws', () => {
    // Suppress console.error for this test
    const consoleSpy = vi.spyOn(console, 'error').mockImplementation(() => {})

    render(
      <ErrorBoundary>
        <ThrowError shouldThrow={true} />
      </ErrorBoundary>,
    )

    expect(screen.getByText('Something went wrong')).toBeInTheDocument()
    expect(
      screen.getByText(/The app encountered an unexpected error/),
    ).toBeInTheDocument()

    // Should show error details
    expect(screen.getByText(/Test error message/)).toBeInTheDocument()

    // Should show action buttons
    expect(screen.getByRole('button', { name: /reload/i })).toBeInTheDocument()
    expect(screen.getByRole('button', { name: /dismiss/i })).toBeInTheDocument()
    expect(
      screen.getByRole('button', { name: /copy details/i }),
    ).toBeInTheDocument()
    expect(
      screen.getByRole('link', { name: /report issue/i }),
    ).toBeInTheDocument()

    consoleSpy.mockRestore()
  })

  it('should match snapshot for error state', () => {
    const consoleSpy = vi.spyOn(console, 'error').mockImplementation(() => {})

    const { container } = render(
      <ErrorBoundary>
        <ThrowError shouldThrow={true} />
      </ErrorBoundary>,
    )

    expect(container.firstChild).toMatchSnapshot()

    consoleSpy.mockRestore()
  })

  it('should match snapshot for normal state', () => {
    const { container } = render(
      <ErrorBoundary>
        <div data-testid='child-content'>Test content</div>
      </ErrorBoundary>,
    )

    expect(container.firstChild).toMatchSnapshot()
  })
})
