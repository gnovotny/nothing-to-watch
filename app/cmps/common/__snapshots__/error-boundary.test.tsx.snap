// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ErrorBoundary > should match snapshot for error state 1`] = `
<div
  class="fixed inset-0 z-50 grid h-dvh w-dvw place-items-center bg-background p-6 text-foreground"
>
  <div
    class="max-w-2xl rounded-lg border border-border bg-card p-6 shadow-sm"
  >
    <h1
      class="mb-2 font-bold text-xl"
    >
      Something went wrong
    </h1>
    <p
      class="mb-4 text-muted-foreground text-sm"
    >
      The app encountered an unexpected error. This can happen if your browser or GPU doesnâ€™t fully support WebGL or required extensions.
    </p>
    <div
      class="mb-4 rounded-md border border-border bg-popover p-3 text-xs"
    >
      <p
        class="mb-1 font-semibold"
      >
        What you can try
      </p>
      <ul
        class="list-inside list-disc space-y-1"
      >
        <li>
          Reload this page
        </li>
        <li>
          Update your browser/graphics drivers
        </li>
        <li>
          Try a different browser (Chrome/Firefox)
        </li>
      </ul>
    </div>
    <pre
      class="mb-4 max-h-48 overflow-auto rounded-md border border-border bg-muted p-3 text-[11px] leading-tight"
    >
      message: Test error message
stack: Error: Test error message
    at ThrowError (/home/g/projects/private/nothing-to-watch/app/cmps/common/error-boundary.test.tsx:22:11)
    at Object.react-stack-bottom-frame (/home/g/projects/private/nothing-to-watch/node_modules/react-dom/cjs/react-dom-client.development.js:23863:20)
    at renderWithHooks (/home/g/projects/private/nothing-to-watch/node_modules/react-dom/cjs/react-dom-client.development.js:5529:22)
    at updateFunctionComponent (/home/g/projects/private/nothing-to-watch/node_modules/react-dom/cjs/react-dom-client.development.js:8897:19)
    at beginWork (/home/g/projects/private/nothing-to-watch/node_modules/react-dom/cjs/react-dom-client.development.js:10522:18)
    at runWithFiberInDEV (/home/g/projects/private/nothing-to-watch/node_modules/react-dom/cjs/react-dom-client.development.js:1522:13)
    at performUnitOfWork (/home/g/projects/private/nothing-to-watch/node_modules/react-dom/cjs/react-dom-client.development.js:15140:22)
    at workLoopSync (/home/g/projects/private/nothing-to-watch/node_modules/react-dom/cjs/react-dom-client.development.js:14956:41)
    at renderRootSync (/home/g/projects/private/nothing-to-watch/node_modules/react-dom/cjs/react-dom-client.development.js:14936:11)
    at performWorkOnRoot (/home/g/projects/private/nothing-to-watch/node_modules/react-dom/cjs/react-dom-client.development.js:14462:44)
userAgent: Mozilla/5.0 (linux) AppleWebKit/537.36 (KHTML, like Gecko) jsdom/26.1.0
    </pre>
    <div
      class="flex flex-wrap gap-2"
    >
      <button
        class="rounded-md border border-input bg-secondary px-3 py-2 text-sm hover:bg-secondary/80"
        type="button"
      >
        Reload
      </button>
      <button
        class="rounded-md border border-input bg-secondary px-3 py-2 text-sm hover:bg-secondary/80"
        type="button"
      >
        Copy details
      </button>
      <a
        class="rounded-md border border-input bg-secondary px-3 py-2 text-sm hover:bg-secondary/80"
        href="https://github.com/example/repo"
        rel="noreferrer noopener"
        target="_blank"
      >
        Report issue
      </a>
      <button
        class="rounded-md border border-input bg-secondary px-3 py-2 text-sm hover:bg-secondary/80"
        type="button"
      >
        Dismiss
      </button>
    </div>
  </div>
</div>
`;

exports[`ErrorBoundary > should match snapshot for normal state 1`] = `
<div
  data-testid="child-content"
>
  Test content
</div>
`;
